shoop/campaigns/migrations/0006_campaign_effects.py

```diff
@@ -0,0 +1,197 @@
+                ('id', models.AutoField(primary_key=True, auto_created=True, verbose_name='ID', serialize=False)),
+            ],
+            options={
+                'abstract': False,
+            },
+        ),
+        migrations.RemoveField(
+            model_name='basketcampaign',
+            name='discount_amount_value',
+        ),
+        migrations.RemoveField(
+            model_name='basketcampaign',
+            name='discount_percentage',
+        ),
+        migrations.RemoveField(
+            model_name='catalogcampaign',
+            name='discount_amount_value',
+        ),
+        migrations.RemoveField(
+            model_name='catalogcampaign',
+            name='discount_percentage',
+        ),
+        migrations.CreateModel(
+            name='DiscountAmount',
+            fields=[
+                ('campaigneffect_ptr', models.OneToOneField(to='campaigns.CampaignEffect', serialize=False, parent_link=True, auto_created=True, primary_key=True)),
+                ('discount_amount_value', shoop.core.fields.MoneyValueField(default=None, max_digits=36, null=True, help_text='Flat amount of discount.', decimal_places=9, blank=True, verbose_name='discount amount value')),
+            ],
+            options={
+                'abstract': False,
+            },
+            bases=('campaigns.campaigneffect',),
+        ),
+        migrations.CreateModel(
+            name='DiscountPercentage',
+            fields=[
+                ('campaigneffect_ptr', models.OneToOneField(to='campaigns.CampaignEffect', serialize=False, parent_link=True, auto_created=True, primary_key=True)),
+                ('discount_percentage', models.DecimalField(max_digits=6, null=True, help_text='The discount percentage for this campaign.', decimal_places=5, blank=True, verbose_name='discount percentage')),
+            ],
+            options={
+                'abstract': False,
+            },
+            bases=('campaigns.campaigneffect',),
+        ),
+        migrations.AddField(
+            model_name='campaigneffect',
+            name='polymorphic_ctype',
+            field=models.ForeignKey(to='contenttypes.ContentType', editable=False, null=True, related_name='polymorphic_campaigns.campaigneffect_set+'),
+        ),
+        migrations.AddField(
+            model_name='basketcampaign',
+            name='effects',
+            field=models.ManyToManyField(to='campaigns.CampaignEffect', related_name='basket_campaigns', blank=True, verbose_name='effects', help_text='When campaign is triggered, these effects give end result.'),
+        ),
+        migrations.AddField(
+            model_name='catalogcampaign',
+            name='effects',
+            field=models.ManyToManyField(to='campaigns.CampaignEffect', related_name='catalog_campaigns', blank=True, verbose_name='effects', help_text='When campaign is triggered, these effects give end result.'),
+        ),
+        migrations.RunPython(add_effects, cache_effects)
```

Comment:
Why don't you just move the RemoveFields to bottom and do the migration in db just after everything new is created but before the fields are removed? That way you wouldn't need to story anything to memory and there should be no risk of losing data if migration is aborted.